controllers/nstemplateset/space_roles.go

```diff
@@ -0,0 +1,116 @@
+
+	"github.com/go-logr/logr"
+)
+
+type spaceRolesManager struct {
+	*statusManager
+}
+
+// ensure ensures that the space roles for the users exist.
+// Returns `true, nil` if something was changed, `false, nil` if nothing changed, `false, err` if an error occurred
+func (r *spaceRolesManager) ensure(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet) (bool, error) {
+	logger = logger.WithValues("nstemplateset_name", nsTmplSet.Name)
+
+	nss, err := fetchNamespacesByOwner(r.Client, nsTmplSet.Name)
+	if err != nil {
+		return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err,
+			"failed to list namespaces for workspace '%s'", nsTmplSet.Name)
+	}
+	logger.Info("ensuring space roles", "namespace_count", len(nss), "role_count", len(nsTmplSet.Spec.SpaceRoles))
+	changed := false
+	for _, ns := range nss {
+		// space roles previously applied
+		// read annotation to see what was applied last time, so we can compare with the new SpaceRoles and remove all outdated resources (based on their kind/names)
+		lastAppliedSpaceRoles := []toolchainv1alpha1.NSTemplateSetSpaceRole{}
+		if currentSpaceRolesAnnotation, exists := ns.Annotations[toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey]; exists && currentSpaceRolesAnnotation != "" {
+			if err := json.Unmarshal([]byte(currentSpaceRolesAnnotation), &lastAppliedSpaceRoles); err != nil {
+				return false, errors.Wrap(err, "unable to decode current space roles in annotation")
+			}
+		}
+		// compare last-applied vs spec to see if there's anything obsolete to do
+		if reflect.DeepEqual(nsTmplSet.Spec.SpaceRoles, lastAppliedSpaceRoles) {
+			logger.Info("no space role to update", "namespace", ns.Name)
+			continue
+		}
+		lastAppliedSpaceRoleObjs, err := r.getSpaceRolesObjects(&ns, lastAppliedSpaceRoles)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdateForSpaceRolesFailure(logger, nsTmplSet, err, "failed to retrieve last applied space roles")
+		}
+		// space roles to apply now
+		spaceRoleObjs, err := r.getSpaceRolesObjects(&ns, nsTmplSet.Spec.SpaceRoles)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdateForSpaceRolesFailure(logger, nsTmplSet, err, "failed to retrieve space roles to apply")
+		}
+
+		deleted, err := deleteObsoleteObjects(logger, r.Client, lastAppliedSpaceRoleObjs, spaceRoleObjs)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusUpdateFailed, err, "failed to delete redundant objects in namespace '%s'", ns.Name)
+		}
+		changed = changed || deleted
+		// labels to apply on all new objects
+		var labels = map[string]string{
+			toolchainv1alpha1.ProviderLabelKey: toolchainv1alpha1.ProviderLabelValue,
+			toolchainv1alpha1.OwnerLabelKey:    nsTmplSet.GetName(),
+		}
+		logger.Info("creating space role objects")
+		for _, obj := range spaceRoleObjs {
+			logger.Info(fmt.Sprintf("creating/updating %s/%s", obj.GetNamespace(), obj.GetName()))
+		}
```

Comment:
do we really need that? wouldn't it be better to add some logging mechanism to the ApplyClient?

Reply:
removed in 3d00213. We can add them in toolchain-common later if needed.